### ISSUE TITLE
Enhancement - Integrated GIF and emojis

### ISSUE THREAD

**Issue by EdiWeeks** [damus-io/damus#590]
Integrated GIF and emojis

**Comment by alltheseas** [comment]
https://tenor.com/gifapi/documentation

https://developers.giphy.com/

**Comment by alltheseas** [comment]
https://plebhy.com/api @scoder1747 🙏

**Comment by alltheseas** [comment]
@fishcakeday if damus builds a nostr.build GIF keyboard, is it somehow possible to filter the GIFs that pop up according to the npub's respective WOT in order to reduce chance the user does not see GIFs the dont want to see?

cc @ericholguin

**Comment by alltheseas** [comment]
update from @fishcakeday relevant to moderation/nsfw considerations for a GIF keyboard

>  now all new uploads will get the moderation header, ex.: https://i.nostr.build/UmRQB45Kn2FTe2kP.webp

> x-moderation:
SAFE:0.998,QUESTIONABLE:0.000,UNSAFE:0.001
x-original-content-sha256:
92b2a0ecbaf3a34c3129273d045628a45da4ed8cc7b20ac6c9e192a3c5cc3383
x-owner-npub:
npub137c5pd8gmhhe0njtsgwjgunc5xjr2vmzvglkgqs5sjeh972gqqxqjak37w

> this will cover, images, gifs, and videos. Since we cache at the edge, it maybe possible that headers are not present until cache expires or they are updated (20-60 seconds after upload) - so please ensure that it is not always expected. it is normalized between 1 & 0, so x-moderation:
SAFE:0.998,QUESTIONABLE:0.000,UNSAFE:0.001 means that we are 99% sure it is SAFE, QUESTIONABLE is something we are not sure about 100% but maybe unsafe, unsafe usually accurate but can catch a few images/videos that are safe

update from @fishcakeday for media associated with npubs:

> and also, we have been publishing x-owner-npub for a while now, but only for the authenticated uploads. if you need to use it in anything

**Comment by alltheseas** [comment]
@fishcakeday if @ericholguin were to add a search box to the nostr.build gifs, and a user typed e.g. "cat" would only gifs with cats remain?

**Comment by alltheseas** [comment]
Let me check with the team

**Comment by alltheseas** [comment]
@happylemonprogramming want to help @ericholguin

**Comment by alltheseas** [comment]
> @alltheseas sorry I missed this message earlier. Not yet, we still need to index the GIFs and implement the search API. Please share your timeline for the implementation, so I can prioritize accordingly.

@fishcakeday damus has gif keyboard nearly there, and would like to test with a search box! I think this nb gif + search will get folks excited. 

We'd be interested in having it for the upcoming damus testflight, and release to appstore this quarter.

Might yall be open to adding search in the near term?

**Comment by happylemonprogramming** [comment]
Whoa! I just saw this, sorry.

The impression I got from @jb55 was that relying on APIs like Tenor/GIPHY
and even solely Nostr.Build were a no-go. That's why I built GifBuddy
because he wanted to sort through NIP94 events. Now we have over 3,000
NIP94 events and you can search through them using the NIP94 standard,
here's a Python implementation:
https://github.com/believethehype/gifbuddy/blob/main/nostrgifsearch.py

I have that code currently running and working on GifBuddy now where it
searches through these NIP94 events and returns gifs (like you'd expect
from your gif keyboard):
https://gifbuddy.lol/nostr

Let me know how else I can help!
That would be cool to get this integrated!

One downside, and I may be biased here, is that if people use the Damus GIF
search function then we won't be building out the NIP94 GIF library like
GifBuddy is doing. Hopefully that makes sense to everyone, but I can
elaborate where needed.

Thanks,
Lemon

**Comment by jb55** [comment]
yeah I definitely prefer the nip94 approach and would like to see that
as the backend for our gif keyboard

**Comment by alltheseas** [comment]
@happylemonprogramming join damus dot io slash devchat

**Comment by happylemonprogramming** [comment]
> > @alltheseas sorry I missed this message earlier. Not yet, we still need to index the GIFs and implement the search API. Please share your timeline for the implementation, so I can prioritize accordingly.
> 
> @fishcakeday damus has gif keyboard nearly there, and would like to test with a search box! I think this nb gif + search will get folks excited.
> 
> We'd be interested in having it for the upcoming damus testflight, and release to appstore this quarter.
> 
> Might yall be open to adding search in the near term?

Where's the best place to start to see the code you have so far on the gif keyboard functionality? Do you already have the UI elements that you want and just need the backend GIF database? Because I have a lot of that already.

Gifbuddy currently queries relays for Kind 1063 events every 2 minutes and stores that into a local database so that the results are loaded instantly

Point me in the right direction and I'll get started

**Comment by alltheseas** [comment]
nostrdb API query needs to be exposed

- @jb55 

Front end is ready. 

Swift back end needed. 
API tenor key needed as fallback

**Comment by alltheseas** [comment]
Official meme added 

![image](https://github.com/user-attachments/assets/bda38386-4f77-4e51-a905-0586307146e2)

https://damus.io/nevent1qqsz937a3d5cfjrjyz6v95kykfl3tymxfany9ldwgl2grdjswjdlxzcu33y3q

**Comment by alltheseas** [comment]
@happylemonprogramming progress 

https://damus.io/note1hy00e75shpjf2dvq28mpfwnzr34wnp66hxd6xrmpv52ku3tdc2tqvz7m3q
